### PR TITLE
feat: Support Basic Auto Completion

### PR DESCRIPTION
**Pull request by CsCherrYY**
Parts of #956
completion

This PR try to get local gradle library file via the current vscode-java settings, and parse the gradle library file to get the content.

For method calls, we'll calculate call the method calls from the root to the inner calls, and use "org.gradle.api.Project" to match them one by one, for the final calls, we get the related Gradle Library and offer the completion results.

### CODE REVIEW

**Comment by Eskibear** [comment]
in the screenshot above, when selecting `gradleApi()`, why it just inserts "gradleApi" without parenthesis?

**Comment by CsCherrYY** [comment]
> in the screenshot above, when selecting `gradleApi()`, why it just inserts "gradleApi" without parenthesis?

It's a bug. Just try in IntelliJ and find it's behavior:
- for those methods have only one method and it's arugment is `groovy.lang.Closure`, offers snippet `{}`
- otherwise, offers `()`

I'd like to change to this in the next commit.

File: gradle-language-server/build.gradle [hunk 1]
```diff
@@ -16,6 +16,7 @@ dependencies {
   implementation "org.eclipse.lsp4j:org.eclipse.lsp4j.jsonrpc:0.12.0"
   implementation "org.codehaus.groovy:groovy-eclipse-batch:3.0.8-01"
   implementation "com.google.code.gson:gson:2.8.7"
+  implementation "org.apache.bcel:bcel:6.5.0"
```
Comment (Eskibear) [inline gradle-language-server/build.gradle:19]: if it's shipped together, you should update `thirdpartynotice`

File: gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java [hunk 2]
```diff
@@ -42,4 +53,14 @@ public static String getStringBeforePosition(String text, Range range, Position
     }
     return text.substring(0, position.getCharacter() - start.getCharacter());
   }
+
+  public static void applySetting(GradleServices services, Object settings) {
```
Comment (Eskibear) [inline gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java]: `applySetting` anything to do with LSP?

File: gradle-language-server/build.gradle (hunk 1 above)
Comment (CsCherrYY) [reply gradle-language-server/build.gradle:19]: should we update the `thirdpartynotice` for other dependencies?

File: gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java (hunk 2 above)
Comment (CsCherrYY) [reply gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java]: The setting values are converted from LSP Json objects. Maybe something like `GradleServivesUtils` is a better place for this method?

File: gradle-language-server/build.gradle (hunk 1 above)
Comment (Eskibear) [reply gradle-language-server/build.gradle:19]: Yes, we'd better do that.

File: gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java (hunk 2 above)
Comment (Eskibear) [reply gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java]: to me, LSPUtils should host utilities like converting between LSP objects and other data structures.

Comment (Eskibear) [reply gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java]: can it be a public member of class `GradleServices`?

Comment (CsCherrYY) [reply gradle-language-server/src/main/java/com/microsoft/gradle/utils/LSPUtils.java]: > can it be a public member of class `GradleServices`?

Sure.

File: gradle-language-server/build.gradle (hunk 1 above)
Comment (CsCherrYY) [reply gradle-language-server/build.gradle:19]: track in #976

File: gradle-language-server/src/main/java/com/microsoft/gradle/GradleServices.java [hunk 3]
```diff
@@ -131,7 +131,17 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
   @Override
   public void didChangeConfiguration(DidChangeConfigurationParams params) {
     Map<?, ?> settings = new Gson().fromJson((JsonElement) params.getSettings(), Map.class);
-    LSPUtils.applySetting(this, settings);
+    this.applySetting(this, settings);
+  }
+
+  public void applySetting(GradleServices services, Object settings) {
```
Comment (Eskibear) [inline gradle-language-server/src/main/java/com/microsoft/gradle/GradleServices.java]: why do you need to pass a `GradleService`? Can it be `public void applySetting(Object settings)` ?

**Review by Eskibear — APPROVED** [review]

**Pull request merged**